File: crates/chthl/src/token.rs
//! This module defines the tokens that make up the CHTL language.

#[derive(Debug, PartialEq, Clone)]
pub enum Token {
    // Special tokens
    Illegal(String), // Represents a token that is not valid
    Eof,             // End of file

    // Identifiers and literals
    Ident(String),  // e.g., div, color, 100px, 50. Also used for unquoted literals.
    String(String), // "a string" or 'a string'
    Number(String), // e.g., 50, 1.5

    // Delimiters
    LBrace,     // {
    RBrace,     // }
    LParen,     // (
    RParen,     // )
    LBracket,   // [
    RBracket,   // ]
    Semicolon,  // ;
    Comma,      // ,
    Colon,      // :
    Assign,     // =

    // Special symbols
    At,         // @
    Hash,       // #
    Dot,        // .
    Ampersand,  // &
    Question,   // ?
    Arrow,      // ->

    // Operators for conditional expressions (to be expanded)
    Gt,             // >
    Lt,             // <
    LogicalAnd,     // &&
    LogicalOr,      // ||
    Bang,           // !
    Plus,           // +
    Minus,          // -
    Asterisk,       // *
    Slash,          // /

    // Comments
    GeneratorComment(String), // -- a comment
}

File: pkg/lexer/token.go
package lexer

type TokenType string

type Token struct {
	Type     TokenType
	Literal  string
	Position int // The starting position of the token in the input string
}

const (
	ILLEGAL TokenType = "ILLEGAL"
	EOF     TokenType = "EOF"
	IDENT   TokenType = "IDENT"
	STRING  TokenType = "STRING"
	NUMBER  TokenType = "NUMBER"
	ASSIGN  TokenType = "="
	COLON   TokenType = ":"
	PLUS    TokenType = "+"
	MINUS   TokenType = "-"
	BANG    TokenType = "!"
	ASTERISK TokenType = "*"
	SLASH    TokenType = "/"
	LBRACE    TokenType = "{"
	RBRACE    TokenType = "}"
	LBRACKET  TokenType = "["
	RBRACKET  TokenType = "]"
	AT        TokenType = "@"
	SEMICOLON TokenType = ";"
	DOT       TokenType = "."
	HASH      TokenType = "#"
	AMPERSAND TokenType = "&"
	FROM      TokenType = "from"
	AS        TokenType = "as"
	COMMENT     TokenType = "COMMENT"
	GEN_COMMENT TokenType = "GEN_COMMENT"
)

var keywords = map[string]TokenType{
	"from": FROM,
	"as":   AS,
}

func LookupIdent(ident string) TokenType {
	if tok, ok := keywords[ident]; ok {
		return tok
	}
	return IDENT
}
